src/app/clusters/dishwasher-alarm-server/dishwasher-alarm-server.cpp

```diff
@@ -0,0 +1,299 @@
+}
+
+EmberAfStatus DishwasherAlarmServer::GetStateValue(EndpointId endpoint, BitMask<AlarmMap> * state)
+{
+    EmberAfStatus status = Attributes::State::Get(endpoint, state);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: reading state, err:0x%x", status);
+        return status;
+    }
+
+    ChipLogProgress(Zcl, "Dishwasher Alarm: State ep%d value: %" PRIu32 "", endpoint, state->Raw());
+
+    return status;
+}
+
+EmberAfStatus DishwasherAlarmServer::SetMaskValue(EndpointId endpoint, const BitMask<AlarmMap> mask)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    status               = Attributes::Mask::Set(endpoint, mask);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: writing  mask, err:0x%x", status);
+        return status;
+    }
+
+    ChipLogProgress(Zcl, "Dishwasher Alarm: Mask ep%d value: %" PRIu32 "", endpoint, mask.Raw());
+
+    // Whenever there is change in Mask, State should change accordingly.
+    BitMask<AlarmMap> state;
+    status = GetStateValue(endpoint, &state);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return status;
+    }
+
+    if (state != (mask & state))
+    {
+        state  = mask & state;
+        status = SetStateValue(endpoint, state);
+    }
+
+    return status;
+}
+
+EmberAfStatus DishwasherAlarmServer::SetLatchValue(EndpointId endpoint, const BitMask<AlarmMap> latch)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    status               = Attributes::Latch::Set(endpoint, latch);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: writing  latch, err:0x%x", status);
+        return status;
+    }
+
+    ChipLogProgress(Zcl, "Dishwasher Alarm: latch ep%d value: %" PRIu32 "", endpoint, latch.Raw());
+    return status;
+}
+
+EmberAfStatus DishwasherAlarmServer::SetStateValue(EndpointId endpoint, BitMask<AlarmMap> newState)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    BitMask<AlarmMap> currentState;
+
+    status = Attributes::State::Get(endpoint, &currentState);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: reading  state, err:0x%x", status);
+        return status;
+    }
+
+    status = Attributes::State::Set(endpoint, newState);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: writing  state, err:0x%x", status);
+        return status;
+    }
+
+    ChipLogProgress(Zcl, "Dishwasher Alarm: State ep%d value: %" PRIu32 "", endpoint, newState.Raw());
+
+    // Maintain consistency between latch and state states
+    BitMask<AlarmMap> latch;
+    status = GetLatchValue(endpoint, &latch);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return status;
+    }
+
+    if (newState != latch)
+    {
+        latch.SetField(latch, newState.Raw());
```

Comment:
This logic is also not making sense.  What is this trying to do?

At this point: this needs tests to go with the code before it can merge.  Because as far as I can tell, this code is not tested at all, manually or otherwise, and catching all the problems by code inspection is really painful and error-prone.

Reply:
Sure, I add a unit test now.